kaggle-classification/keras-trainer/model.py

```diff
@@ -0,0 +1,184 @@
+from keras.layers import Conv1D
+from keras.layers import Dense
+from keras.layers import Dropout
+from keras.layers import Embedding
+from keras.layers import Flatten
+from keras.layers import Input
+from keras.layers import MaxPooling1D
+from keras.layers import Activation
+from keras.layers import Concatenate
+from keras.models import load_model
+from keras.models import Model
+from keras.preprocessing.sequence import pad_sequences
+from keras.preprocessing.text import Tokenizer
+from os.path import expanduser
+from sklearn import metrics
+
+FLAGS = None
+
+DEFAULT_HPARAMS = tf.contrib.training.HParams(
+    learning_rate=0.00005,
+    dropout_rate=0.3,
+    batch_size=128,
+    epochs=5,
+    max_sequence_length=250,
+    embedding_dim=100)
+
+LABELS = ['toxic', 'severe_toxic', 'obscene', 'threat', 'insult', 'identity_hate']
+
+
+class AttentionToxModel():
+  """Toxicity model using CNN + Attention"""
+  
+  def __init__(self,
+               model_path,
+               embeddings_path,
+               hparams=DEFAULT_HPARAMS):
+    self.model_path = model_path
+    self.embeddings_path = embeddings_path
+    self.hparams = hparams
+    print("Setting up tokenizer...")
+    self.tokenizer = self._setup_tokenizer()
+    print("Setting up embedding matrix...")
+    self.embedding_matrix = self._setup_embedding_matrix()
+    print("Loading model...")
+    self._load_model()
+  
+  def train(self, train):
+    model = self._build_model()
+    train_comment = self._prep_texts(train['comment_text'])
+    train_labels = [train[label] for label in LABELS]
+
+    callbacks = [
+        ModelCheckpoint(
+            self.model_path, save_best_only=True, verbose=True),
+        EarlyStopping(
+            monitor='val_loss', mode='auto')
+    ]
+
+    model.fit(
+        x=train_comment, y=train_labels,
+        batch_size=self.hparams.batch_size,
+        epochs=self.hparams.epochs,
+        validation_split=0.1,
+        callbacks=callbacks)
+    
+    self._load_model()
+
+  def predict(self, texts):
+    data = self._prep_texts(texts)
+    return self.model.predict(data)
+
+  def score_auc(self, data):
+    predictions = self.predict(data['comment_text'])
+    scores = []
+    for idx, label in enumerate(LABELS):
+      labels = np.array(data['toxic'])
+      score = metrics.roc_auc_score(labels, predictions[idx].flatten())
+      scores.append(score)
+      print("{} has AUC {}".format(label, score))
+    print("Avg AUC {}".format(np.mean(scores)))
+
+  def _prep_texts(self, texts):
+    return pad_sequences(self.tokenizer.texts_to_sequences(texts), maxlen=self.hparams.max_sequence_length)
+
+  def _load_model(self):
+    try:
+      self.model = load_model(self.model_path)
+      print('Model loaded from: {}'.format(self.model_path))
+    except IOError:
+      print('Could not load model at: {}'.format(self.model_path))
+
+  def _setup_tokenizer(self):
+    words = []
+    with open(self.embeddings_path) as f:
```

Comment:
Is there a way to avoid reading this file twice?

Reply:
Let me defer. I don't think it's simple because the tokenizer needs to be fully setup because we created the embedding matrix.